portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -20,92 +20,218 @@
+                response.sendRedirect("/");
+                return;
+            }
+        }
+        if (maxNumberOfComments < 0 || maxNumberOfComments > MAX_COMMENTS_NUMBER) {
+            response.sendRedirect("/");
+            return;
+        }
+
+        // get lastTimestamp parameter
+        String lastTimestampString = request.getParameter("timestamp");
+        long lastTimestamp;
+        if (lastTimestampString == null) {
+            lastTimestamp = System.currentTimeMillis() + 10;
+        } else {
+            try {
+                lastTimestamp = Long.parseLong(lastTimestampString);
+            } catch (NumberFormatException e) {
+                response.sendRedirect("/");
+                return;
+            }
+        }
+
+        // get direction parameter
+        String direction = request.getParameter("direction");
+        if (direction == null) {
+            direction = "next";
+        }
+        if (!direction.equals("next") && !direction.equals("previous")) {
+            response.sendRedirect("/");
+            return;
+        }
+
+        // get commentsOnPage parameter (on current page)
+        String commentsOnPageString = request.getParameter("commentsonpage");
+        int commentsOnPage;
+        if (commentsOnPageString == null) {
+            commentsOnPage = maxNumberOfComments;
+        } else {
+            try {
+                commentsOnPage = Integer.parseInt(commentsOnPageString);
+            } catch (NumberFormatException e) {
+                response.sendRedirect("/");
+                return;
+            }
+        }
+
+        // query options
+        Filter timeFilter;
+        SortDirection sortDirection;
+        FetchOptions fetchOptions;
+
+        // set query options depending on whether the user want to see next or previous comments
+        if (direction.equals("next")) {
+            timeFilter =
+                    new FilterPredicate("timestamp", FilterOperator.LESS_THAN, lastTimestamp);
+            sortDirection = SortDirection.DESCENDING;
+            fetchOptions = FetchOptions.Builder.withLimit(maxNumberOfComments);
+        } else {
+            // if previous - sort query in other direction and get comments that are
+            // currently on page and those we will load
+            timeFilter =
+                    new FilterPredicate("timestamp", FilterOperator.GREATER_THAN_OR_EQUAL, lastTimestamp);
+            sortDirection = SortDirection.ASCENDING;
+            fetchOptions = FetchOptions.Builder.withLimit(maxNumberOfComments + commentsOnPage);
+        }
+
+        // get datastore
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        // check if user wants next comments, and current comment is already the last one
+        // if so, return current comments
+        long minTimestamp = getMinTimestamp(datastore);
+        if (minTimestamp == lastTimestamp && direction.equals("next")) {
```

Comment:
=== comparison

Reply:
=== is for Javascript and the code is in Java so I think it's fine the way it is

Reply:
My bad, got confused, thanks for pointing this out.